File: structural-design-pattern/composite-pattern/composite.ts
/**
 * Composite Pattern
 * =================
 * Category: Structural Design Pattern
 *
 * Intent:
 *   Compose objects into tree structures to represent part-whole hierarchies.
 *   Clients treat individual objects (leaves) and compositions (branches) uniformly.
 *
 * When to use:
 *   - Tree/hierarchical structures (files/folders, org charts, UI components)
 *   - When clients should treat single objects and groups the same way
 *   - When operations should propagate recursively through the tree
 *
 * Key Participants:
 *   - Component (interface): Common interface for leaves and composites
 *   - Leaf: End node with no children
 *   - Composite: Node that contains children
 */

// ---------------------------------------------------------------------------
// Component — common interface
// ---------------------------------------------------------------------------
interface FileSystemEntry {
  name: string;
  size(): number;
  display(indent?: number): string;
  search(keyword: string): string[];
}

// ---------------------------------------------------------------------------
// Leaf — a file (no children)
// ---------------------------------------------------------------------------
class File implements FileSystemEntry {
  constructor(public name: string, private sizeBytes: number) {}

  size(): number {
    return this.sizeBytes;
  }

  display(indent: number = 0): string {
    const prefix = "  ".repeat(indent);
    return `${prefix}F ${this.name} (${this.sizeBytes} bytes)`;
  }

  search(keyword: string): string[] {
    return this.name.toLowerCase().includes(keyword.toLowerCase())
      ? [this.name]
      : [];
  }
}

// ---------------------------------------------------------------------------
// Composite — a directory (has children)
// ---------------------------------------------------------------------------
class Directory implements FileSystemEntry {
  private children: FileSystemEntry[] = [];

  constructor(public name: string) {}

  add(entry: FileSystemEntry): this {
    this.children.push(entry);
    return this;
  }

  remove(name: string): void {
    this.children = this.children.filter((c) => c.name !== name);
  }

  size(): number {
    // Recursively sums all children — uniform treatment
    return this.children.reduce((sum, child) => sum + child.size(), 0);
  }

  display(indent: number = 0): string {
    const prefix = "  ".repeat(indent);
    const lines = [`${prefix}D ${this.name}/ (${this.size()} bytes)`];
    for (const child of this.children) {
      lines.push(child.display(indent + 1));
    }
    return lines.join("\n");
  }

  search(keyword: string): string[] {
    const results: string[] = [];
    if (this.name.toLowerCase().includes(keyword.toLowerCase())) {
      results.push(this.name + "/");
    }
    for (const child of this.children) {
      results.push(...child.search(keyword));
    }
    return results;
  }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
const root = new Directory("project");

const src = new Directory("src");
src.add(new File("main.ts", 2048));
src.add(new File("utils.ts", 1024));
src.add(new File("config.ts", 512));

const tests = new Directory("tests");
tests.add(new File("test_main.ts", 1536));
tests.add(new File("test_utils.ts", 768));

const docs = new Directory("docs");
docs.add(new File("README.md", 4096));
docs.add(new File("CHANGELOG.md", 2048));

root.add(src).add(tests).add(docs).add(new File(".gitignore", 128));

console.log(root.display());
console.log(`\nTotal project size: ${root.size()} bytes`);
console.log(`Source code size: ${src.size()} bytes`);
console.log(`\nSearch for 'test': ${root.search("test")}`);
console.log(`Search for 'main': ${root.search("main")}`);

File: structural-design-pattern/composite-pattern/composite.py
"""
Composite Pattern
=================
Category: Structural Design Pattern

Intent:
    Compose objects into tree structures to represent part-whole hierarchies.
    Lets clients treat individual objects (leaves) and compositions (branches)
    uniformly through a common interface.

When to use:
    - When you have a tree/hierarchical structure (files/folders, org charts, UI components)
    - When clients should treat single objects and groups the same way
    - When operations should propagate recursively through the tree

Key Participants:
    - Component: Common interface for both leaves and composites
    - Leaf: End node with no children (e.g., a File)
    - Composite: Node that contains children (e.g., a Directory)
"""

from abc import ABC, abstractmethod


# ---------------------------------------------------------------------------
# Component — common interface
# ---------------------------------------------------------------------------
class FileSystemEntry(ABC):
    """Every entry in the file system — file or directory — shares this interface."""

    def __init__(self, name: str) -> None:
        self.name = name

    @abstractmethod
    def size(self) -> int:
        """Returns the size in bytes."""
        ...

    @abstractmethod
    def display(self, indent: int = 0) -> str:
        """Returns a tree-like string representation."""
        ...

    @abstractmethod
    def search(self, keyword: str) -> list[str]:
        """Find entries whose name contains the keyword."""
        ...


# ---------------------------------------------------------------------------
# Leaf — a file (no children)
# ---------------------------------------------------------------------------
class File(FileSystemEntry):
    def __init__(self, name: str, size_bytes: int) -> None:
        super().__init__(name)
        self._size = size_bytes

    def size(self) -> int:
        return self._size

    def display(self, indent: int = 0) -> str:
        prefix = "  " * indent
        return f"{prefix}📄 {self.name} ({self._size} bytes)"

    def search(self, keyword: str) -> list[str]:
        if keyword.lower() in self.name.lower():
            return [self.name]
        return []


# ---------------------------------------------------------------------------
# Composite — a directory (has children)
# ---------------------------------------------------------------------------
class Directory(FileSystemEntry):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self._children: list[FileSystemEntry] = []

    def add(self, entry: FileSystemEntry) -> "Directory":
        """Add a child (file or subdirectory). Returns self for chaining."""
        self._children.append(entry)
        return self

    def remove(self, name: str) -> None:
        self._children = [c for c in self._children if c.name != name]

    def size(self) -> int:
        # Recursively sums sizes of all children — uniform treatment
        return sum(child.size() for child in self._children)

    def display(self, indent: int = 0) -> str:
        prefix = "  " * indent
        lines = [f"{prefix}📁 {self.name}/ ({self.size()} bytes)"]
        for child in self._children:
            lines.append(child.display(indent + 1))
        return "\n".join(lines)

    def search(self, keyword: str) -> list[str]:
        results: list[str] = []
        if keyword.lower() in self.name.lower():
            results.append(self.name + "/")
        for child in self._children:
            results.extend(child.search(keyword))
        return results


# ---------------------------------------------------------------------------
# Usage Example
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    # Build a file system tree
    root = Directory("project")

    src = Directory("src")
    src.add(File("main.py", 2048))
    src.add(File("utils.py", 1024))
    src.add(File("config.py", 512))

    tests = Directory("tests")
    tests.add(File("test_main.py", 1536))
    tests.add(File("test_utils.py", 768))

    docs = Directory("docs")
    docs.add(File("README.md", 4096))
    docs.add(File("CHANGELOG.md", 2048))

    root.add(src)
    root.add(tests)
    root.add(docs)
    root.add(File(".gitignore", 128))

    # Display the tree — works uniformly for files and directories
    print(root.display())

    # Size calculation propagates through the tree
    print(f"\nTotal project size: {root.size()} bytes")
    print(f"Source code size: {src.size()} bytes")

    # Search works recursively across the entire tree
    print(f"\nSearch for 'test': {root.search('test')}")
    print(f"Search for 'main': {root.search('main')}")
